### DISCUSSION TITLE
why bindMethod don't work

### DISCUSSION THREAD

**Discussion by AntistressStore** [laravel/framework#50216] [General]
Hello. I need to introduce a second argument to the events subscriber class. I do this via app->bindMethod as recommended. 

```php
// in provider
class SportApiParserServiceProvider extends ServiceProvider implements DeferrableProvider
{
public function register(): void
{
    $this->app->bindMethod(
            [SportEventScheduleSubscriber::class, 'handleSportEventScheduleCreated'],
            function (SportEventScheduleSubscriber $subs, Application $app) {
                return $subs->handleSportEventScheduleCreated(
                         $app->make(SportEventScheduleCreated::class), 
                         $app->make(FolderService::class));
            }
        );

}

public function provides(): array
    {
        return [
            SportEventSchedule::class,
            SportEventScheduleSubscriber::class,
        ];
    }
}

// in events subscriber

class SportEventScheduleSubscriber
{
    /**
     * Create the event listener.
     */
    public function __construct()
    {
        //
    }

    public function handleSportEventScheduleCreated(SportEventScheduleCreated $event, FolderService $folderService): void
    {
        Cache::flush();

        $folderService->reorganizeBkEventsUsingAction();

        DefaultLog::logSuccess('jobs', $this->successMessage($event->settings));
    }
}

```

However, the application throws an error.

Too few arguments to function SportEventScheduleSubscriber::handleSportEventScheduleCreated(), 1 passed in /var/www/e/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php on line 478 and exactly 2 expected SportEventScheduleSubscriber.php

and does not pass 2 argument.
I tried moving bindMethod to boot method of my App\Providers\AppServiceProvider service provider.

And it doesn't work either. 

When I debug the application, I can see that it does not go to the method binding point in the provider.
If i pass FolderService $folderService to the class constructor SportEventScheduleSubscriber and use $this everything will work. But I only need this class for a single method, not in the constructor for the whole class. How can I make the bindMethod work?
How to fix the problem?

**Comment by henzeb** [comment]
I don't think there's binding support at all on those methods on a subscriber on events. if it was, it would have tried to bind that folderService instance and never complained about a thing. 

i also think you have misunderstood how events work. How does the event get dispatched here?

**Comment by AntistressStore** [comment]
The fact is that there is no description of bindMethod in the Service Container manual for version 10.
https://laravel.com/docs/10.x/container

The only mention of bindMethod is at https://laravel.com/docs/10.x/queues#handle-method-dependency-injection
It notes that it is advisable to implement it in the boot method of your App\Providers\AppServiceProvider service provider. And there is no clarification on whether if you use your service provider or use bindMethod in the register method of your service provider, it will work.
bind, singletone and other methods work great through the register method in the service provider.

Accordingly, I have a question - to what extent is it a working concept at the moment to use bindMethod? And if this is a working concept, why doesn't it work in this case? What needs to be done to make it work?

**Reply by henzeb** [reply]
the handle method is injected using the App container by default, but before it does it, it checks if there is a binding to the method of that particular Job, and if so, then uses that to inject into the method.

Your method on your subscriber isn't injected by the framework at all.

What you need to do is to add a property to your event and when you dispatch that event add the `FolderService` to it, OR add the folder service as dependency in your subscribers constructor....

**Reply by AntistressStore** [reply]
Thank you for your advice.  I'll apply your advice.
